Ingress/controllers/nginx-third-party/controller.go

```diff
@@ -0,0 +1,316 @@
+		UpdateFunc: func(old, cur interface{}) {
+			if !reflect.DeepEqual(old, cur) {
+				glog.V(2).Infof("nginx rc changed, syncing")
+			}
+			lbc.configQueue.enqueue(cur)
+		},
+	}
+
+	lbc.configLister.Store, lbc.configController = framework.NewInformer(
+		&cache.ListWatch{
+			WatchFunc: func(options api.ListOptions) (watch.Interface, error) {
+				return kubeClient.
+					Get().
+					Prefix("watch").
+					Namespace(lbInfo.RCNamespace).
+					Resource("resource").
+					Name(lbInfo.RCName).
+					Watch()
+			},
+		},
+		&api.ReplicationController{}, resyncPeriod, configHandlers)
+
+	return &lbc, nil
+}
+
+func ingressListFunc(c *client.Client, ns string) func() (runtime.Object, error) {
+	return func() (runtime.Object, error) {
+		return c.Extensions().Ingress(ns).List(labels.Everything(), fields.Everything())
+	}
+}
+
+func ingressWatchFunc(c *client.Client, ns string) func(options api.ListOptions) (watch.Interface, error) {
+	return func(options api.ListOptions) (watch.Interface, error) {
+		return c.Extensions().Ingress(ns).Watch(
+			labels.Everything(), fields.Everything(), options)
+	}
+}
+
+// syncIngress manages Ingress create/updates/deletes.
+func (lbc *loadBalancerController) syncIngress(key string) {
+	glog.V(2).Infof("Syncing Ingress %v", key)
+
+	obj, ingExists, err := lbc.ingLister.Store.GetByKey(key)
+	if err != nil {
+		lbc.ingQueue.requeue(key, err)
+		return
+	}
+
+	if !ingExists {
+		return
+	}
+
+	// No need to reload nginx
+	// Instead we invoke a POST /update-ingress to update the Ingress rules to use
+	ingList := lbc.ingLister.Store.List()
+	lbc.ngx.SyncIngress(ingList)
+
+	ing := *obj.(*extensions.Ingress)
+	if lbc.updateIngressStatus(ing); err != nil {
+		lbc.recorder.Eventf(&ing, "Status", "%v", err)
+		lbc.ingQueue.requeue(key, err)
+	}
+	return
+}
+
+// syncIngress manages Ingress create/updates/deletes.
+func (lbc *loadBalancerController) syncConfig(key string) {
+	glog.V(2).Infof("Syncing config %v", key)
+
+	obj, configExists, err := lbc.configLister.Store.GetByKey(key)
+	if err != nil {
+		lbc.configQueue.requeue(key, err)
+		return
+	}
+
+	if !configExists {
+		return
+	}
+
+	rc := *obj.(*api.ReplicationController)
+	ngxCfgAnn, _ := annotations(rc.Annotations).getNginxConfig()
+	tcpSvcAnn, _ := annotations(rc.Annotations).getNginxConfig()
+
+	ngxConfig, err := lbc.ngx.ReadConfig(ngxCfgAnn)
+	if err != nil {
+		glog.Warning("%v", err)
+	}
+
+	// TODO: tcp services can change (new item in the annotation list)
+	tcpServices := getTcpServices(lbc.client, tcpSvcAnn)
+	lbc.ngx.Reload(ngxConfig, tcpServices)
+
+	return
+}
+
+// updateIngressStatus updates the IP and annotations of a loadbalancer.
+// The annotations are parsed by kubectl describe.
+func (lbc *loadBalancerController) updateIngressStatus(ing extensions.Ingress) error {
+	ingClient := lbc.client.Extensions().Ingress(ing.Namespace)
+
+	ip := lbc.lbInfo.PodIP
+	currIng, err := ingClient.Get(ing.Name)
```

Comment:
please add a todo to skip a get everytime. I don't think it's a big deal for ingress controller, but we do this in other places because usually clients have qps limits and if you're the only thing updating the ingress you don't need to reget unless the update actually fails.